File: saga/services.go
package saga

import (
	"fmt"
	"log/slog"
)

type InventoryService struct {
	bus           *EventBus
	logger        *slog.Logger
	failForOrders map[string]bool
	reservations  map[string]int
}

func NewInventoryService(bus *EventBus, logger *slog.Logger, failForOrders map[string]bool) *InventoryService {
	if failForOrders == nil {
		failForOrders = map[string]bool{}
	}
	s := &InventoryService{bus: bus, logger: logger, failForOrders: failForOrders, reservations: map[string]int{}}
	bus.Subscribe("inventory.reserve.requested", s.Reserve)
	bus.Subscribe("inventory.release.requested", s.Release)
	return s
}

func (s *InventoryService) Reserve(event Event) {
	orderID := event.Payload["order_id"].(string)
	qty := event.Payload["quantity"].(int)

	if s.failForOrders[orderID] {
		err := fmt.Sprintf("insufficient stock for order %s", orderID)
		s.logger.Error("inventory_reserve_failed", "saga_id", event.SagaID, "error", err)
		s.bus.MustPublish(Event{Name: "inventory.reserve.failed", SagaID: event.SagaID, Payload: map[string]any{"order_id": orderID, "error": err}})
		return
	}

	s.reservations[orderID] += qty
	s.bus.MustPublish(Event{Name: "inventory.reserved", SagaID: event.SagaID, Payload: map[string]any{"order_id": orderID, "quantity": qty}})
}

func (s *InventoryService) Release(event Event) {
	orderID := event.Payload["order_id"].(string)
	qty := s.reservations[orderID]
	delete(s.reservations, orderID)
	s.bus.MustPublish(Event{Name: "inventory.released", SagaID: event.SagaID, Payload: map[string]any{"order_id": orderID, "quantity": qty}})
}

type PaymentService struct {
	bus           *EventBus
	logger        *slog.Logger
	failForOrders map[string]bool
	charges       map[string]float64
}

func NewPaymentService(bus *EventBus, logger *slog.Logger, failForOrders map[string]bool) *PaymentService {
	if failForOrders == nil {
		failForOrders = map[string]bool{}
	}
	s := &PaymentService{bus: bus, logger: logger, failForOrders: failForOrders, charges: map[string]float64{}}
	bus.Subscribe("payment.charge.requested", s.Charge)
	bus.Subscribe("payment.refund.requested", s.Refund)
	return s
}

func (s *PaymentService) Charge(event Event) {
	orderID := event.Payload["order_id"].(string)
	amount := event.Payload["amount"].(float64)

	if s.failForOrders[orderID] {
		err := fmt.Sprintf("card declined for order %s", orderID)
		s.logger.Error("payment_charge_failed", "saga_id", event.SagaID, "error", err)
		s.bus.MustPublish(Event{Name: "payment.charge.failed", SagaID: event.SagaID, Payload: map[string]any{"order_id": orderID, "error": err}})
		return
	}

	s.charges[orderID] = amount
	s.bus.MustPublish(Event{Name: "payment.charged", SagaID: event.SagaID, Payload: map[string]any{"order_id": orderID, "amount": amount}})
}

func (s *PaymentService) Refund(event Event) {
	orderID := event.Payload["order_id"].(string)
	amount := s.charges[orderID]
	delete(s.charges, orderID)
	s.bus.MustPublish(Event{Name: "payment.refunded", SagaID: event.SagaID, Payload: map[string]any{"order_id": orderID, "amount": amount}})
}

type ShippingService struct {
	bus           *EventBus
	logger        *slog.Logger
	failForOrders map[string]bool
	shipments     map[string]string
}

func NewShippingService(bus *EventBus, logger *slog.Logger, failForOrders map[string]bool) *ShippingService {
	if failForOrders == nil {
		failForOrders = map[string]bool{}
	}
	s := &ShippingService{bus: bus, logger: logger, failForOrders: failForOrders, shipments: map[string]string{}}
	bus.Subscribe("shipping.create.requested", s.Create)
	return s
}

func (s *ShippingService) Create(event Event) {
	orderID := event.Payload["order_id"].(string)
	address := event.Payload["address"].(string)

	if s.failForOrders[orderID] {
		err := fmt.Sprintf("shipping unavailable for order %s", orderID)
		s.logger.Error("shipping_create_failed", "saga_id", event.SagaID, "error", err)
		s.bus.MustPublish(Event{Name: "shipping.create.failed", SagaID: event.SagaID, Payload: map[string]any{"order_id": orderID, "error": err}})
		return
	}

	trackingID := fmt.Sprintf("TRK-%s", orderID)
	s.shipments[orderID] = address
	s.bus.MustPublish(Event{Name: "shipping.created", SagaID: event.SagaID, Payload: map[string]any{"order_id": orderID, "tracking_id": trackingID}})
}

File: saga/services.py
import logging
from typing import Dict, Set

from saga.event_bus import EventBus
from saga.models import Event


class InventoryService:
    def __init__(self, bus: EventBus, logger: logging.Logger, fail_for_orders: Set[str] | None = None) -> None:
        self.bus = bus
        self.logger = logger
        self.fail_for_orders = fail_for_orders or set()
        self.reservations: Dict[str, int] = {}

        self.bus.subscribe("inventory.reserve.requested", self.reserve)
        self.bus.subscribe("inventory.release.requested", self.release)

    def reserve(self, event: Event) -> None:
        order_id = event.payload["order_id"]
        qty = event.payload["quantity"]

        self.logger.info("inventory_reserve_started", extra={"saga_id": event.saga_id, "order_id": order_id, "qty": qty})

        if order_id in self.fail_for_orders:
            error = f"Estoque insuficiente para pedido {order_id}"
            self.logger.error("inventory_reserve_failed", extra={"saga_id": event.saga_id, "error": error})
            self.bus.publish(Event("inventory.reserve.failed", event.saga_id, {"order_id": order_id, "error": error}))
            return

        self.reservations[order_id] = self.reservations.get(order_id, 0) + qty
        self.logger.info("inventory_reserved", extra={"saga_id": event.saga_id, "order_id": order_id, "qty": qty})
        self.bus.publish(Event("inventory.reserved", event.saga_id, {"order_id": order_id, "quantity": qty}))

    def release(self, event: Event) -> None:
        order_id = event.payload["order_id"]
        qty = self.reservations.pop(order_id, 0)
        self.logger.warning("inventory_released", extra={"saga_id": event.saga_id, "order_id": order_id, "qty": qty})
        self.bus.publish(Event("inventory.released", event.saga_id, {"order_id": order_id, "quantity": qty}))


class PaymentService:
    def __init__(self, bus: EventBus, logger: logging.Logger, fail_for_orders: Set[str] | None = None) -> None:
        self.bus = bus
        self.logger = logger
        self.fail_for_orders = fail_for_orders or set()
        self.charges: Dict[str, float] = {}

        self.bus.subscribe("payment.charge.requested", self.charge)
        self.bus.subscribe("payment.refund.requested", self.refund)

    def charge(self, event: Event) -> None:
        order_id = event.payload["order_id"]
        amount = event.payload["amount"]

        self.logger.info("payment_charge_started", extra={"saga_id": event.saga_id, "order_id": order_id, "amount": amount})

        if order_id in self.fail_for_orders:
            error = f"Cartão recusado para pedido {order_id}"
            self.logger.error("payment_charge_failed", extra={"saga_id": event.saga_id, "error": error})
            self.bus.publish(Event("payment.charge.failed", event.saga_id, {"order_id": order_id, "error": error}))
            return

        self.charges[order_id] = amount
        self.logger.info("payment_charged", extra={"saga_id": event.saga_id, "order_id": order_id, "amount": amount})
        self.bus.publish(Event("payment.charged", event.saga_id, {"order_id": order_id, "amount": amount}))

    def refund(self, event: Event) -> None:
        order_id = event.payload["order_id"]
        amount = self.charges.pop(order_id, 0.0)
        self.logger.warning("payment_refunded", extra={"saga_id": event.saga_id, "order_id": order_id, "amount": amount})
        self.bus.publish(Event("payment.refunded", event.saga_id, {"order_id": order_id, "amount": amount}))


class ShippingService:
    def __init__(self, bus: EventBus, logger: logging.Logger, fail_for_orders: Set[str] | None = None) -> None:
        self.bus = bus
        self.logger = logger
        self.fail_for_orders = fail_for_orders or set()
        self.shipments: Dict[str, str] = {}

        self.bus.subscribe("shipping.create.requested", self.create)

    def create(self, event: Event) -> None:
        order_id = event.payload["order_id"]
        address = event.payload["address"]

        self.logger.info("shipping_create_started", extra={"saga_id": event.saga_id, "order_id": order_id, "address": address})

        if order_id in self.fail_for_orders:
            error = f"Transportadora indisponível para pedido {order_id}"
            self.logger.error("shipping_create_failed", extra={"saga_id": event.saga_id, "error": error})
            self.bus.publish(Event("shipping.create.failed", event.saga_id, {"order_id": order_id, "error": error}))
            return

        tracking_id = f"TRK-{order_id}"
        self.shipments[order_id] = tracking_id
        self.logger.info("shipping_created", extra={"saga_id": event.saga_id, "order_id": order_id, "tracking_id": tracking_id})
        self.bus.publish(Event("shipping.created", event.saga_id, {"order_id": order_id, "tracking_id": tracking_id}))
